experimental/transliterator_parser/src/compile/pass2.rs

```diff
@@ -0,0 +1,379 @@
+
+impl<'a, 'p> Pass2<'a, 'p> {
+    pub(super) fn run(
+        result: DirectedPass1Result<'p>,
+        var_definitions: &'a HashMap<String, &'p [parse::Element]>,
+    ) -> Result<ds::RuleBasedTransliterator<'static>> {
+        let mut pass2 = Self::try_new(result.data.counts, var_definitions)?;
+        pass2.compile(result.groups, result.filter)
+    }
+
+    fn try_new(
+        counts: SpecialConstructCounts,
+        var_definitions: &'a HashMap<String, &'p [parse::Element]>,
+    ) -> Result<Self> {
+        Ok(Pass2 {
+            var_table: MutVarTable::try_new_from_counts(counts)?,
+            var_definitions,
+            var_to_char: HashMap::new(),
+        })
+    }
+
+    fn compile(
+        &mut self,
+        rule_groups: super::RuleGroups<'p>,
+        global_filter: Option<UnicodeSet>,
+    ) -> Result<ds::RuleBasedTransliterator<'static>> {
+        let mut compiled_transform_groups: Vec<VarZeroVec<'static, ds::SimpleIdULE>> = Vec::new();
+        let mut compiled_conversion_groups: Vec<VarZeroVec<'static, ds::RuleULE>> = Vec::new();
+
+        for (transform_group, conversion_group) in rule_groups {
+            let compiled_transform_group: Vec<_> = transform_group
+                .into_iter()
+                .map(|id| self.compile_single_id(&id))
+                .collect();
+            compiled_transform_groups.push(VarZeroVec::from(&compiled_transform_group));
+
+            let compiled_conversion_group: Vec<_> = conversion_group
+                .into_iter()
+                .map(|rule| self.compile_conversion_rule(rule))
+                .collect();
+            compiled_conversion_groups.push(VarZeroVec::from(&compiled_conversion_group));
+        }
+
+        let res = ds::RuleBasedTransliterator {
+            visibility: true, // TODO(#3736): use metadata
+            filter: global_filter
+                .map(|f| f.code_points().clone())
+                .unwrap_or(CodePointInversionList::all()),
+            id_group_list: VarZeroVec::from(&compiled_transform_groups),
+            rule_group_list: VarZeroVec::from(&compiled_conversion_groups),
+            variable_table: self.var_table.finalize(),
+        };
+
+        Ok(res)
+    }
+
+    fn compile_conversion_rule(&mut self, rule: UniConversionRule<'p>) -> ds::Rule<'static> {
+        let ante = self.compile_section(rule.ante, parse::ElementLocation::Source);
+        let key = self.compile_section(rule.key, parse::ElementLocation::Source);
+        let post = self.compile_section(rule.post, parse::ElementLocation::Source);
+        let replacer = self.compile_section(rule.replacement, parse::ElementLocation::Target);
+        let cursor_offset = rule.cursor_offset;
+        ds::Rule {
+            ante: ante.into(),
+            key: key.into(),
+            post: post.into(),
+            replacer: replacer.into(),
+            cursor_offset,
+        }
+    }
+
+    fn compile_single_id(&mut self, id: &parse::SingleId) -> ds::SimpleId<'static> {
```

Comment:
this doesn't use `self`, so it doesn't need to be a method. Maybe put it on `parse::SingleId`?

Reply:
I prefer not to spread out the compilation over multiple modules, would an associated function here be fine as well?

Reply:
Discussed: associated function fine. Doing in https://github.com/unicode-org/icu4x/pull/3827